qiskit/passmanager/base_tasks.py

```diff
@@ -0,0 +1,227 @@
+                log_msg = f"Pass: {self.name()} - {running_time * 1000:.5f} (ms)"
+                logger.info(log_msg)
+                if callback is not None:
+                    callback(
+                        task=self,
+                        passmanager_ir=ret,
+                        property_set=self.property_set,
+                        running_time=running_time,
+                        count=status.count,
+                    )
+            self.update_status(status, run_state)
+        return ret
+
+    def update_status(
+        self,
+        status: WorkflowStatus,
+        run_state: RunState,
+    ):
+        """Update workflow status.
+
+        Args:
+            status: Status object to mutably update.
+            run_state: Completion status of current task.
+        """
+        status.previous_run = run_state
+        if run_state == RunState.SUCCESS:
+            status.count += 1
+            status.completed_passes.add(self)
+
+    @abstractmethod
+    def run(
+        self,
+        passmanager_ir: PassManagerIR,
+    ) -> PassManagerIR:
+        """Run optimization task.
+
+        Args:
+            passmanager_ir: Qiskit IR to optimize.
+
+        Returns:
+            Optimized Qiskit IR.
+        """
+        pass
+
+
+class BaseController(Task, ABC):
+    """Base class of controller.
+
+    A controller is built with a collection of pass manager tasks,
+    and a subclass provides a custom logic to choose next task to run.
+    Note a controller can be nested into another controller,
+    and controller itself doesn't provide any subroutine to modify the input IR.
+    """
+
+    def __init__(
+        self,
+        options: dict[str, Any] | None = None,
+    ):
+        """Create new flow controller.
+
+        Args:
+            options: Option for this flow controller.
+        """
+        self._options = options or {}
+
+    @abstractmethod
+    def iter_tasks(
+        self,
+        property_set: PropertySet,
+    ) -> Generator[Task]:
+        """A custom logic to choose a next task to run.
+
+        Controller subclass can consume the property set to build a proper task pipeline.
+        This indicates the order of task execution is only determined at running time.
+
+        Args:
+            property_set: A mutable data collection shared among all tasks.
+
+        Yields:
+            Next task to run.
+        """
+        pass
+
+    def execute(
+        self,
+        passmanager_ir: PassManagerIR,
+        status: WorkflowStatus | None = None,
+        property_set: PropertySet | None = None,
+        callback: Callable = None,
+    ) -> PassManagerIR:
+        # Overriding this method is not safe.
+        # Pass subclass must keep current implementation.
+        # Especially, task execution may break when method signature is modified.
+
+        if property_set is None:
+            property_set = PropertySet()
+        if status is None:
+            status = WorkflowStatus()
```

Comment:
Do we really need to handle defaults for `status` and `property_set` here?  Can we simplify things for downstream implementers by making default construction the sole responsibility of `PassManager`?

Reply:
No, updated in [d889ac0](https://github.com/Qiskit/qiskit/pull/10127/commits/d889ac02c24db7374866d547e9ec2ea173b68512)